Approximate entropy

kahaaga opened the pull request:
## What is this PR?

This PR is an implementation of the approximate entropy (ApEn) algorithm from Pincus (1991), a widely used measure of the regularity of time series (for a summary, see Delgado-Bonal & Marshak, 2019).

Note: the sample entropy (PR #71) is a proposed improvement of the approximate entropy.

## Implementation

A simple tree search is used to locate within-radius-`r` neighbors. During testing, I found this approach to be more than an order of magnitude faster, both in terms of runtime and allocations, than using a naive loop-based approach.

## Tests

Tests are generic, as the original paper lack short, concise examples to test on. However, in the doc example, I reproduce the Henon map example in Pincus (1991) with reasonable accuracy, so I think the implementation should be correct.

## References

Pincus, S. M. (1991). Approximate entropy as a measure of system complexity. Proceedings of the National Academy of Sciences, 88(6), 2297-2301.

Delgado-Bonal, A., & Marshak, A. (2019). Approximate entropy and sample entropy: A comprehensive tutorial. Entropy, 21(6), 541.

Datseris commented on src/approximate_entropy/approximate_entropy.jl:
```diff
@@ -0,0 +1,80 @@
+using Neighborhood
+using StatsBase
+using Distances
+
+export approx_entropy
+
```
I'm confused here. Why is this like that instead of following the established API that defines some kind of entropy estimator and then calls the entropy function?

kahaaga replied:
I fixed this is the latest commit. The PR now defines the `ApproximateEntropy` estimator which works with `genentropy` (but not `probabilities`). The `approx_entropy` function is still included (as for permutation entropy, approximate entropy is widely used, so I figured a convenience function is nice to have).

Datseris commented on src/complexity_measures/approximate_entropy.jl:
```diff
@@ -0,0 +1,91 @@
+export approx_entropy
+
+# So we don't need to import StatsBase
+_std(x) = sqrt(sum((x .- mean(x)) .^ 2) / (length(x) - 1))
```
Just `using Statistics`...

kahaaga replied:
> Just using Statistics...

Or that, yes :D

Datseris commented:
Shouldn't there be a small `## Description` section in the docstring? For the tests, you could use what you already have in the example plot; compare the value with the value from the table, i.e., that it is close to 0.394 for N = 3000.

Datseris commented on docs/src/examples.md:
```diff
@@ -225,4 +225,82 @@ end
+using Entropies, DynamicalSystems, CairoMakie
+
+# Equation 13 in Pincus (1991)
+function eom_henon(x, p, n)
+    R = p[1]
+    x, y = (x...,)
```
```suggestion
function eom_henon(u, p, n)
    R = p[1]
    x, y = u
```

kahaaga commented:
@Datseris I've now added:

- Fixed your comments.
- Improved notation and added an algorithm description.
- A test based on Pincus' example. I removed the MATLAB tests, because it turns out they do not (according to their documentation) adjust the normalization constant for embedding lags != 1, and they use `<` instead of `<=` for neighbor searches. Thus, results would not be comparable in general. 
- Added `test/Project.toml`, so we don't have to make DynamicalSystemsBase.jl a dependency for just installing Entropies.jl.  Using a `Project.toml` file is anyways the standard for Julia 1.3 an onwards.

kahaaga commented:
This has now been updated to use the new complexity api (#134). I've also implemented changes based on comments from #71.

Datseris commented on docs/src/complexity_measures.md:
````diff
@@ -11,3 +11,17 @@ complexity_normalized
+ApproxEntropy
+```
+
+## Convenience functions
+
+We provide a few convenience functions for widely used "entropy-like" complexity measures, such as "approximate entropy". Other arbitrary specialized convenience functions can easily be defined in a couple lines of code.
+
+```@docs
+approx_entropy
+```
````
Well now I am confused. I thought we agreed that we do this table thingy and therefore do not provide convenience functions anymore.

kahaaga replied:
> Well now I am confused. I thought we agreed that we do this table thingy and therefore do not provide convenience functions anymore.

We agreed on having exactly two convenience functions for complexity measures too, for educational purposes, like we do for entropy. Those would be sample entropy and approximate entropy, because they are the most widely used ones. Citing [yourself](https://github.com/JuliaDynamics/Entropies.jl/issues/133#issuecomment-1282262636) in #134: 

> Okay let's keep the sample_entropy and approxiamte_entropy as the only "convenience complexity functions".

Remaining complexity measures go in the table.

Datseris replied:
I made this comment before the existence of the table was on the table however (pun not intended). If we have a table, which satisfies the "search engine" argument, do we need these convenience methods? Or, we can have a docs section that discusses the library design principles. I have one in Agents.jl as well. Okay. Leave the "convenience method" here. Once we have the table, I'll do a PR that proposes my section.

Datseris commented on docs/src/complexity_measures.md:
````diff
@@ -11,3 +11,17 @@ complexity_normalized
+## Approximate entropy
+
+```@docs
+ApproxEntropy
````
Can we name this `ApproximateEntropy` directly? I'ts already long eniough where saving the extra 5 characters doesn't matter.

kahaaga replied:
Yes, why not.
